transpiler/transpiler.go

```diff
@@ -0,0 +1,562 @@
+// is no operating system. A strict C++ subset (without RTTI) is used.
+//
+// ** Do not take this project too seriously **
+//
+// Challenges and transformations:
+//  - function with multiple return values is converted to returning as a
+//    temporary struct
+//  - string constant is converted to const char *
+//  - interface{} is converted to void *
+//  - interface inheritance is figured out at parsing time
+//  - out of bound check for slice and strings
+//  - string indexing is done via byte offset, not runes
+//  - struct are manually zero initialized
+//  - recursive type resolution of imported packages
+//
+// Out of scope:
+//  - channel type
+//  - defer statement
+//  - go statement
+//  - len function
+//  - map type
+//  - range expression
+//  - select statement
+//  - switch statement
+//  - anonymous function
+//  - function pointer
+//  - pointer to member of struct
+//  - unnamed struct embedding
+//  - memory management, all memory allocation is leaked
+//  - dynamic type casting involving RTTI
+//  - using the STL in the generated code
+package transpiler
+
+import (
+	"bytes"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"io"
+	"io/ioutil"
+	"reflect"
+	"strings"
+
+	"github.com/kr/pretty"
+)
+
+// Transpile converts a Go (.go) source file to C++ (.cc).
+func Transpile(out io.Writer, in io.Reader) (*ast.File, error) {
+	fset := token.NewFileSet()
+	// Keep a copy of the input file to do a byte offset to line conversion.
+	content, err := ioutil.ReadAll(in)
+	if err != nil {
+		return nil, err
+	}
+	f, err := parser.ParseFile(fset, "src.go", bytes.NewReader(content), parser.ParseComments)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse: %s", err)
+	}
+	lines := make([]int, 0, 128)
+	for i, c := range content {
+		if c == '\n' {
+			lines = append(lines, i)
+		}
+	}
+	o := &output{out, content, lines, nil, f.Comments, nil}
+	/*
+		// Explicitly push everything up to package name so it doesn't get printed.
+		// It's kind of annoying as there's no PackageStmt so it has to be explicitly
+		// emulated.
+		if f.Package != 0 && len(lines) != 0 {
+			o.out.Write(content[:f.Package-1])
+			o.lastNode = &fakeNode{token.Pos(lines[o.findLine(int(f.Package))-1] + 1)}
+		}
+	*/
+	for _, i := range f.Imports {
+		o.Writef(i, "")
+	}
+	for _, d := range f.Decls {
+		if err := handleDecl(o, d); err != nil {
+			return f, err
+		}
+		if o.err != nil {
+			break
+		}
+	}
+	return f, o.err
+}
+
+type fakeNode struct {
+	e token.Pos
+}
+
+func (f *fakeNode) Pos() token.Pos {
+	return 0
+}
+
+func (f *fakeNode) End() token.Pos {
+	return f.e
+}
+
+type output struct {
+	out      io.Writer
+	content  []byte
+	lines    []int
+	lastNode ast.Node
+	c        []*ast.CommentGroup
+	err      error
+}
+
+// Writef makes sure that all comments up to the point where the Node is
+// declared are flushed.
+func (o *output) Writef(n ast.Node, format string, a ...interface{}) {
```

Comment:
I wonder if we should just handle comments like we handle any other ast node, rather than special casing it on every write.

That would allow us to have `output` just be a thin wrapper over `io.Writer` and `fmt.Fprintf` directly to it.